File: Trinity.Runtime/platform_src/android/app/src/main/java/org/elastos/trinity/runtime/passwordmanager/passwordinfo/BankCardPasswordInfo.java
package org.elastos.trinity.runtime.passwordmanager.passwordinfo;

import org.json.JSONException;
import org.json.JSONObject;

public class BankCardPasswordInfo extends PasswordInfo {
    private BankCardType cardType = null;
    private String accountOwner = null;
    private String cardNumber = null;
    private String expirationDate = null;
    private String cvv = null;
    private String bankName = null;

    public static PasswordInfo fromJsonObject(JSONObject jsonObject) throws Exception {
        BankCardPasswordInfo info = new BankCardPasswordInfo();

        info.fillWithJsonObject(jsonObject);

        return info;
    }

    @Override
    public JSONObject asJsonObject() {
        try {
            JSONObject jsonObject = super.asJsonObject();

            jsonObject.put("cardType", cardType.mValue);
            jsonObject.put("accountOwner", accountOwner);
            jsonObject.put("cardNumber", cardNumber);
            jsonObject.put("expirationDate", expirationDate);
            jsonObject.put("cvv", cvv);
            jsonObject.put("bankName", bankName);

            return jsonObject;
        } catch (
                JSONException e) {
            e.printStackTrace();
            return null;
        }
    }

    @Override
    public void fillWithJsonObject(JSONObject jsonObject) throws Exception {
        // Fill base fields
        super.fillWithJsonObject(jsonObject);

        // Fill specific fields
        if (jsonObject.has("type")) {
            this.cardType = BankCardType.fromValue(jsonObject.getInt("cardType"));
        }
        this.accountOwner = jsonObject.optString("accountOwner");
        this.cardNumber = jsonObject.optString("cardNumber");
        this.expirationDate = jsonObject.optString("expirationDate");
        this.cvv = jsonObject.optString("cvv");
        this.bankName = jsonObject.optString("bankName");
    }
}

File: Trinity.Runtime/platform_src/ios/elastOS/Classes/PasswordManager/BankCardPasswordInfo.swift
/*
* Copyright (c) 2020 Elastos Foundation
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the "Software"), to deal
* in the Software without restriction, including without limitation the rights
* to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
* copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
*
* The above copyright notice and this permission notice shall be included in all
* copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
* SOFTWARE.
*/

import Foundation

public class BankCardPasswordInfo : PasswordInfo {
    private var cardType: BankCardType? = nil
    private var accountOwner: String? = nil
    private var cardNumber: String? = nil
    private var expirationDate: String? = nil
    private var cvv: String? = nil
    private var bankName: String? = nil

    public static func fromJsonObject(_ jsonObject: Dictionary<String, Any>) throws -> PasswordInfo {
        let info = BankCardPasswordInfo()

        try info.fillWithJsonObject(jsonObject)

        return info
    }

    public override func asDictionary() -> Dictionary<String, Any>? {
        if var jsonObject = super.asDictionary() {
            if let cardType = cardType {
                jsonObject["cardType"] = cardType.rawValue
            }
            jsonObject["accountOwner"] = accountOwner
            jsonObject["cardNumber"] = cardNumber
            jsonObject["expirationDate"] = expirationDate
            jsonObject["cvv"] = cvv
            jsonObject["bankName"] = bankName

            return jsonObject
        }
        
        return nil
    }

    public override func fillWithJsonObject(_ jsonObject: Dictionary<String, Any>) throws {
        // Fill base fields
        try super.fillWithJsonObject(jsonObject)

        // Fill specific fields
        if (jsonObject.keys.contains("type")) {
            self.cardType = BankCardType.init(rawValue: jsonObject["cardType"] as! Int)
        }
        self.accountOwner = jsonObject["accountOwner"] as? String
        self.cardNumber = jsonObject["cardNumber"] as? String
        self.expirationDate = jsonObject["expirationDate"] as? String
        self.cvv = jsonObject["cvv"] as? String
        self.bankName = jsonObject["bankName"] as? String
    }
}
